pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -34,15 +62,359 @@ func (s *discovery) GetStats() map[string]interface{} {
+		log.Debugf("couldn't snapshot UDP6 sockets: %v", err)
+	}
+
+	nsInfo := namespaceInfo{
+		socketInodeToPort: make(map[uint64]int),
+	}
+
+	for _, sock := range TCP {
+		if sock.St != tcpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range TCP6 {
+		if sock.St != tcpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+	for _, sock := range UDP6 {
+		if sock.St != udpListen {
+			continue
+		}
+		nsInfo.socketInodeToPort[sock.Inode] = int(sock.LocalPort)
+	}
+
+	if !isRoot {
+		nsInfo.addrs, err = getNsAddresses(path)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &nsInfo, nil
+}
+
+func getRootNsInode(procRoot string) (uint32, error) {
+	ns, err := kernel.GetRootNetNamespace(procRoot)
+	if err != nil {
+		return 0, err
+	}
+	defer ns.Close()
+
+	inode, err := kernel.GetInoForNs(ns)
+	if err != nil {
+		return 0, err
+	}
+
+	return inode, nil
+}
+
+// parsingContext holds temporary context not preserved between invocations of
+// the endpoint.
+type parsingContext struct {
+	procRoot  string
+	rootNs    uint32
+	netNsInfo map[uint32]*namespaceInfo
+}
+
+func (s *discovery) getListener(context parsingContext, pid int32) *model.Listener {
+	proc, err := process.NewProcess(pid)
+	if err != nil {
+		return nil
+	}
+
+	name, err := proc.Name()
+	if err != nil {
+		return nil
+	}
+
+	sockets, err := getSockets(proc)
+	if err != nil {
+		return nil
+	}
+	if sockets == nil {
+		return nil
+	}
+
+	ns, err := kernel.GetNetNsInoFromPid(context.procRoot, int(pid))
+	if err != nil {
+		return nil
+	}
+
+	// The socket and network address information are different for each
+	// network namespace.  Since namespaces can be shared between multiple
+	// processes, we cache them to only parse them once per call to this
+	// function.
+	nsInfo, ok := context.netNsInfo[ns]
+	if !ok {
+		nsInfo, err = getNsInfo(int(pid), ns == context.rootNs)
+		if err != nil {
+			return nil
+		}
+
+		context.netNsInfo[ns] = nsInfo
+	}
+
+	haveListeningSocket := false
+	for _, socket := range sockets {
+		if _, ok := nsInfo.socketInodeToPort[socket]; ok {
+			haveListeningSocket = true
+			break
+		}
+	}
+
+	if !haveListeningSocket {
+		return nil
+	}
+
+	var env []string
+	if cached, ok := s.cache[pid]; ok {
+		env = cached.env
+	} else {
+		env, err = proc.Environ()
+		if err != nil {
+			return nil
+		}
+
+		env = filterEnv(env)
+		s.cache[pid] = cacheData{env: env}
+	}
+
+	cwd, err := proc.Cwd()
+	if err != nil {
+		return nil
+	}
+
+	return &model.Listener{
+		PID:     int(pid),
+		Name:    name,
+		Environ: env,
+		CWD:     cwd,
+	}
+}
+
+func (s *discovery) cleanCache(alivePids map[int32]struct{}) {
+	for pid := range s.cache {
+		if _, alive := alivePids[pid]; alive {
+			continue
+		}
+
+		delete(s.cache, pid)
+	}
```

Comment:
AFAIK, deleting from a map does not free up the memory

Reply:
It should free the memory for the strings, right? I added a comment.